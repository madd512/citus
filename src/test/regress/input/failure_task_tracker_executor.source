SELECT citus.mitmproxy('flow.allow()');

SET citus.task_executor_type TO 'task-tracker';
SET citus.enable_unique_job_ids TO false;

SET citus.shard_count = 2; -- one per worker
SET citus.shard_replication_factor = 2; -- one shard per worker

CREATE TABLE agg_test (id int, val int, flag bool, kind int);
SELECT create_distributed_table('agg_test', 'id');
INSERT INTO agg_test VALUES (1, 1, true, 99), (2, 2, false, 99), (2, 3, true, 88);

-- the correct answer
SELECT bool_or(flag) FROM agg_test;

-- prevent coord from giving any tasks out
SELECT citus.mitmproxy('flow.contains(b"task_tracker_assign_task").killall()');

-- the other worker is still here so we should get the right results
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('flow.contains(b"task_tracker_task_status").killall()');
SELECT bool_or(flag) FROM agg_test;

  -- kill it when the coord tries to pull results from the worker
SELECT citus.mitmproxy('flow.contains(b"TO STDOUT WITH").killall()');
SELECT bool_or(flag) FROM agg_test;

-- TODO: try failing the COPY after some data has been sent

SELECT citus.mitmproxy('flow.contains(b"task_tracker_cleanup_job").killall()');
SELECT bool_or(flag) FROM agg_test;

-- TODO: what if you fail the first task_tracker_task_status, but allow the second one?
SELECT citus.mitmproxy('flow.allow()');
DROP TABLE agg_test;

SET citus.shard_replication_factor = 1;  -- workers contain disjoint subsets of the data

CREATE TABLE agg_test (id int, val int, flag bool, kind int);
SELECT create_distributed_table('agg_test', 'id');
INSERT INTO agg_test VALUES (1, 1, true, 99), (2, 2, false, 99), (2, 3, true, 88);

-- for all of these the query should fail since we can't reach all of the shards it hits

SELECT citus.mitmproxy('flow.contains(b"task_tracker_assign_task").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('flow.contains(b"task_tracker_task_status").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('flow.contains(b"TO STDOUT WITH").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('flow.contains(b"task_tracker_cleanup_job").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('flow.allow()');
DROP TABLE agg_test;

SELECT citus.mitmproxy('flow.allow()');

-- copied over from multi_outer_join.sql

CREATE FUNCTION create_tables() RETURNS void
AS $$
  CREATE TABLE multi_outer_join_left
  (
  	l_custkey integer not null,
  	l_name varchar(25) not null,
  	l_address varchar(40) not null,
  	l_nationkey integer not null,
  	l_phone char(15) not null,
  	l_acctbal decimal(15,2) not null,
  	l_mktsegment char(10) not null,
  	l_comment varchar(117) not null
  );
  
  CREATE TABLE multi_outer_join_right
  (
  	r_custkey integer not null,
  	r_name varchar(25) not null,
  	r_address varchar(40) not null,
  	r_nationkey integer not null,
  	r_phone char(15) not null,
  	r_acctbal decimal(15,2) not null,
  	r_mktsegment char(10) not null,
  	r_comment varchar(117) not null
  );
  COPY multi_outer_join_left FROM '@abs_srcdir@/data/customer-1-10.data' with delimiter '|';
  COPY multi_outer_join_left FROM '@abs_srcdir@/data/customer-11-20.data' with delimiter '|';
  COPY multi_outer_join_right FROM '@abs_srcdir@/data/customer-1-15.data' with delimiter '|';
$$ LANGUAGE SQL;

SELECT create_tables();
SELECT create_distributed_table('multi_outer_join_left', 'l_custkey');
SELECT create_distributed_table('multi_outer_join_right', 'r_custkey');

-- run a simple outer join
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.contains(b"task_tracker_assign_task").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.contains(b"task_tracker_task_status").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.contains(b"TO STDOUT WITH").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.contains(b"task_tracker_cleanup_job").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.allow()');
DROP TABLE multi_outer_join_left;
DROP TABLE multi_outer_join_right;

-- okay, try again with a higher shard replication factor

SET citus.shard_replication_factor = 2; -- one shard per worker
SELECT create_tables();
SELECT create_distributed_table('multi_outer_join_left', 'l_custkey');
SELECT create_distributed_table('multi_outer_join_right', 'r_custkey');

-- the other worker is reachable, so these should work
SELECT citus.mitmproxy('flow.contains(b"task_tracker_assign_task").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.contains(b"task_tracker_task_status").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.contains(b"TO STDOUT WITH").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.contains(b"task_tracker_cleanup_job").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('flow.allow()');
DROP TABLE multi_outer_join_left;
DROP TABLE multi_outer_join_right;
DROP FUNCTION create_tables();
