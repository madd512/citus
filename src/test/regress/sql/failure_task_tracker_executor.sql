SELECT citus.mitmproxy('conn.allow()');

SET citus.task_executor_type TO 'task-tracker';
SET citus.enable_unique_job_ids TO false;

SET citus.shard_count = 2; -- one per worker
SET citus.shard_replication_factor = 2; -- one shard per worker

CREATE TABLE agg_test (id int, val int, flag bool, kind int);
SELECT create_distributed_table('agg_test', 'id');
INSERT INTO agg_test VALUES (1, 1, true, 99), (2, 2, false, 99), (2, 3, true, 88);

-- the correct answer
SELECT bool_or(flag) FROM agg_test;

-- prevent coord from giving any tasks out
SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');

-- the other worker is still here so we should get the right results
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
SELECT bool_or(flag) FROM agg_test;

  -- kill it when the coord tries to pull results from the worker
SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
SELECT bool_or(flag) FROM agg_test;

-- TODO: try failing the COPY after some data has been sent

SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
SELECT bool_or(flag) FROM agg_test;

-- TODO: what if you fail the first task_tracker_task_status, but allow the second one?
SELECT citus.mitmproxy('conn.allow()');
DROP TABLE agg_test;

SET citus.shard_replication_factor = 1;  -- workers contain disjoint subsets of the data

CREATE TABLE agg_test (id int, val int, flag bool, kind int);
SELECT create_distributed_table('agg_test', 'id');
INSERT INTO agg_test VALUES (1, 1, true, 99), (2, 2, false, 99), (2, 3, true, 88);

-- for all of these the query should fail since we can't reach all of the shards it hits

SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
SELECT bool_or(flag) FROM agg_test;

SELECT citus.mitmproxy('conn.allow()');
DROP TABLE agg_test;

SELECT citus.mitmproxy('conn.allow()');

-- copied over from multi_outer_join.sql

CREATE FUNCTION create_tables() RETURNS void
AS $$
  CREATE TABLE multi_outer_join_left ( l_custkey, l_name)
    AS SELECT s.a, to_char(s.a, '999') FROM generate_series(1, 20) AS s(a);
  
  CREATE TABLE multi_outer_join_right ( r_custkey, r_name)
    AS SELECT s.a, to_char(s.a, '999') FROM generate_series(1, 15) AS s(a);
$$ LANGUAGE SQL;

SELECT create_tables();
SELECT create_distributed_table('multi_outer_join_left', 'l_custkey');
SELECT create_distributed_table('multi_outer_join_right', 'r_custkey');

-- run a simple outer join
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.allow()');
DROP TABLE multi_outer_join_left;
DROP TABLE multi_outer_join_right;

-- okay, try again with a higher shard replication factor

SET citus.shard_replication_factor = 2; -- one shard per worker
SELECT create_tables();
SELECT create_distributed_table('multi_outer_join_left', 'l_custkey');
SELECT create_distributed_table('multi_outer_join_right', 'r_custkey');

-- the other worker is reachable, so these should work
SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);

SELECT citus.mitmproxy('conn.allow()');
DROP TABLE multi_outer_join_left;
DROP TABLE multi_outer_join_right;
DROP FUNCTION create_tables();
