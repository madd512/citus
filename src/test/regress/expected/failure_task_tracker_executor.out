SELECT citus.mitmproxy('conn.allow()');
 mitmproxy 
-----------
 
(1 row)

SET citus.task_executor_type TO 'task-tracker';
SET citus.enable_unique_job_ids TO false;
SET citus.shard_count = 2; -- one per worker
SET citus.shard_replication_factor = 2; -- one shard per worker
CREATE TABLE agg_test (id int, val int, flag bool, kind int);
SELECT create_distributed_table('agg_test', 'id');
 create_distributed_table 
--------------------------
 
(1 row)

INSERT INTO agg_test VALUES (1, 1, true, 99), (2, 2, false, 99), (2, 3, true, 88);
-- the correct answer
SELECT bool_or(flag) FROM agg_test;
 bool_or 
---------
 t
(1 row)

-- prevent coord from giving any tasks out
SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');
 mitmproxy 
-----------
 
(1 row)

-- the other worker is still here so we should get the right results
SELECT bool_or(flag) FROM agg_test;
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  could not maintain connection to worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
 bool_or 
---------
 t
(1 row)

SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT bool_or(flag) FROM agg_test;
WARNING:  could not consume data from worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
 bool_or 
---------
 t
(1 row)

  -- kill it when the coord tries to pull results from the worker
SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT bool_or(flag) FROM agg_test;
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  connection not open
CONTEXT:  while executing command on localhost:57640
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 4 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(4)" 
 bool_or 
---------
 t
(1 row)

-- TODO: try failing the COPY after some data has been sent
SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT bool_or(flag) FROM agg_test;
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 5 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(5)" 
 bool_or 
---------
 t
(1 row)

-- TODO: what if you fail the first task_tracker_task_status, but allow the second one?
SELECT citus.mitmproxy('conn.allow()');
 mitmproxy 
-----------
 
(1 row)

DROP TABLE agg_test;
SET citus.shard_replication_factor = 1;  -- workers contain disjoint subsets of the data
CREATE TABLE agg_test (id int, val int, flag bool, kind int);
SELECT create_distributed_table('agg_test', 'id');
 create_distributed_table 
--------------------------
 
(1 row)

INSERT INTO agg_test VALUES (1, 1, true, 99), (2, 2, false, 99), (2, 3, true, 88);
-- for all of these the query should fail since we can't reach all of the shards it hits
SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT bool_or(flag) FROM agg_test;
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  could not maintain connection to worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
ERROR:  failed to execute task 2
SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT bool_or(flag) FROM agg_test;
WARNING:  could not consume data from worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
ERROR:  failed to execute task 2
SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT bool_or(flag) FROM agg_test;
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  connection not open
CONTEXT:  while executing command on localhost:57640
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 8 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(8)" 
ERROR:  failed to execute task 2
SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT bool_or(flag) FROM agg_test;
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 9 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(9)" 
 bool_or 
---------
 t
(1 row)

SELECT citus.mitmproxy('conn.allow()');
 mitmproxy 
-----------
 
(1 row)

DROP TABLE agg_test;
SELECT citus.mitmproxy('conn.allow()');
 mitmproxy 
-----------
 
(1 row)

-- copied over from multi_outer_join.sql
CREATE FUNCTION create_tables() RETURNS void
AS $$
  CREATE TABLE multi_outer_join_left ( l_custkey, l_name)
    AS SELECT s.a, to_char(s.a, '999') FROM generate_series(1, 20) AS s(a);
  
  CREATE TABLE multi_outer_join_right ( r_custkey, r_name)
    AS SELECT s.a, to_char(s.a, '999') FROM generate_series(1, 15) AS s(a);
$$ LANGUAGE SQL;
SELECT create_tables();
 create_tables 
---------------
 
(1 row)

SELECT create_distributed_table('multi_outer_join_left', 'l_custkey');
NOTICE:  Copying data from local table...
 create_distributed_table 
--------------------------
 
(1 row)

SELECT create_distributed_table('multi_outer_join_right', 'r_custkey');
NOTICE:  Copying data from local table...
 create_distributed_table 
--------------------------
 
(1 row)

-- run a simple outer join
SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
 min | max 
-----+-----
   1 |  20
(1 row)

SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  could not maintain connection to worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
ERROR:  failed to execute task 2
SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  could not consume data from worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
ERROR:  failed to execute task 2
SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  connection not open
CONTEXT:  while executing command on localhost:57640
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 13 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(13)" 
ERROR:  failed to execute task 2
SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 14 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(14)" 
 min | max 
-----+-----
   1 |  20
(1 row)

SELECT citus.mitmproxy('conn.allow()');
 mitmproxy 
-----------
 
(1 row)

DROP TABLE multi_outer_join_left;
DROP TABLE multi_outer_join_right;
-- okay, try again with a higher shard replication factor
SET citus.shard_replication_factor = 2; -- one shard per worker
SELECT create_tables();
 create_tables 
---------------
 
(1 row)

SELECT create_distributed_table('multi_outer_join_left', 'l_custkey');
NOTICE:  Copying data from local table...
 create_distributed_table 
--------------------------
 
(1 row)

SELECT create_distributed_table('multi_outer_join_right', 'r_custkey');
NOTICE:  Copying data from local table...
 create_distributed_table 
--------------------------
 
(1 row)

-- the other worker is reachable, so these should work
SELECT citus.mitmproxy('conn.contains(b"task_tracker_assign_task").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  could not maintain connection to worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
 min | max 
-----+-----
   1 |  20
(1 row)

SELECT citus.mitmproxy('conn.contains(b"task_tracker_task_status").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  could not consume data from worker node
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
 min | max 
-----+-----
   1 |  20
(1 row)

SELECT citus.mitmproxy('conn.contains(b"TO STDOUT WITH").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
CONTEXT:  while executing command on localhost:57640
WARNING:  connection not open
CONTEXT:  while executing command on localhost:57640
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  connection error: localhost:57640
DETAIL:  server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 17 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(17)" 
 min | max 
-----+-----
   1 |  20
(1 row)

SELECT citus.mitmproxy('conn.contains(b"task_tracker_cleanup_job").killall()');
 mitmproxy 
-----------
 
(1 row)

SELECT
	min(l_custkey), max(l_custkey)
FROM
	multi_outer_join_left a LEFT JOIN multi_outer_join_right b ON (l_custkey = r_custkey);
WARNING:  could not consume data from worker node
WARNING:  could not receive response for cleanup query result for job 18 on node "localhost:57640" with status 1
HINT:  Manually clean job resources on node "localhost:57640" by running "SELECT task_tracker_cleanup_job(18)" 
 min | max 
-----+-----
   1 |  20
(1 row)

SELECT citus.mitmproxy('conn.allow()');
 mitmproxy 
-----------
 
(1 row)

DROP TABLE multi_outer_join_left;
DROP TABLE multi_outer_join_right;
DROP FUNCTION create_tables();
